### ISSUE TITLE
SEVERE log message writing to closed websocket server

### ISSUE THREAD

**Issue by thheller** [ztellman/aleph#309]
I'm using `aleph` as a websocket server for `shadow-devtools` which provides a websocket CLJS REPL.

Whenever I `(.close http-server)` while a websocket is still connected I get these exceptions in the log. I close `core.async` channels as best I can before calling `.close` but I cannot ensure that they all close fast enough.

Having this marked as `SEVERE` seems like a bit much to me and I don't want to turn off the entire log.

I thought about `:shutdown-executor? false` and using a custom `:executor` since I just want to shut down the http server but not the JVM.

Anything else I can do?

This part does the websocket setup, I might be doing something wrong here:
https://github.com/thheller/shadow-devtools/blob/master/src/main/shadow/cljs/devtools/server/worker/ws.clj#L118-L135

This is with `aleph 0.4.3`.

```
Apr 06, 2017 11:15:16 PM clojure.tools.logging$eval20463$fn__20467 invoke
SEVERE: error in deferred handler
java.util.concurrent.RejectedExecutionException: Executor is shutdown!
	at io.aleph.dirigiste.Executor.execute(Executor.java:300)
	at manifold.deferred.Deferred$fn__21294.invoke(deferred.clj:395)
	at manifold.deferred.Deferred.success(deferred.clj:395)
	at manifold.deferred$success_BANG_.invokeStatic(deferred.clj:243)
	at manifold.deferred$success_BANG_.invoke(deferred.clj:240)
	at manifold.stream.default.Stream$fn__22909.invoke(default.clj:81)
	at manifold.stream.default.Stream.close(default.clj:80)
	at manifold.stream.BufferedStream.close(stream.clj:863)
	at aleph.http.server$websocket_server_handler$reify__26465.channelInactive(server.clj:490)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelInactive(ChannelInboundHandlerAdapter.java:75)
	at io.netty.handler.codec.MessageAggregator.channelInactive(MessageAggregator.java:406)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
	at io.netty.handler.codec.ByteToMessageDecoder.channelInputClosed(ByteToMessageDecoder.java:360)
	at io.netty.handler.codec.ByteToMessageDecoder.channelInactive(ByteToMessageDecoder.java:325)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelInactive(AbstractChannelHandlerContext.java:224)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelInactive(DefaultChannelPipeline.java:1329)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:245)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelInactive(AbstractChannelHandlerContext.java:231)
	at io.netty.channel.DefaultChannelPipeline.fireChannelInactive(DefaultChannelPipeline.java:908)
	at io.netty.channel.AbstractChannel$AbstractUnsafe$7.run(AbstractChannel.java:744)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasksFrom(SingleThreadEventExecutor.java:379)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:354)
	at io.netty.util.concurrent.SingleThreadEventExecutor.confirmShutdown(SingleThreadEventExecutor.java:678)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:452)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144)
	at java.lang.Thread.run(Thread.java:745)

Apr 06, 2017 11:15:16 PM clojure.tools.logging$eval20463$fn__20467 invoke
SEVERE: error in deferred handler
java.util.concurrent.RejectedExecutionException: Executor is shutdown!
	at io.aleph.dirigiste.Executor.execute(Executor.java:300)
	at manifold.deferred.Deferred$fn__21294.invoke(deferred.clj:395)
	at manifold.deferred.Deferred.success(deferred.clj:395)
	at manifold.deferred$success_BANG_.invokeStatic(deferred.clj:243)
	at manifold.deferred$success_BANG_.invoke(deferred.clj:240)
	at manifold.stream.async.CoreAsyncSource$f__23374$fn__23375.invoke(async.clj:44)
	at clojure.core.async.impl.channels.ManyToManyChannel$fn__8909$fn__8910.invoke(channels.clj:95)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by ztellman** [comment]
Yeah, this is a weird timing issue, and certainly shouldn't log as SEVERE.  Similar issues are logged in the test suite for Aleph, so I don't think there's anything wrong with what you're doing.  I need to figure out the right way to fix this, I'll leave the issue open until I come up with a satisfactory solution.

**Comment by thheller** [comment]
I switched to passing in a shared `:executor` and `:shutdown-executor? false` which fixes the problem. Oddly enough even if I `.shutdown` the executor just after calling `.close` on the aleph instance.

**Comment by vspinu** [comment]
This this is not in the stable yet. Just wondering if it's good time for 0.4.4 already.

**Comment by ztellman** [comment]
Definitely past time, sorry.  Just released `0.4.4`.
